0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,146 @@
+
+## Flow
+
+Before the protocol starts, the sender and receiver use a higher-level protocol to agree on the destination account and packet condition.
+
+1. Sender constructs an ILP Prepare packet with the receiver's destination account, the agreed-upon condition, and an amount and expiry of their choice. The sender may include additional data, the formatting of which is determined by the higher-level protocol.
+2. Sender sends the ILP Prepare packet to a connector over some communication channel, such as a Websocket. 
+3. Connector checks whether the sender has sufficient bandwidth (balance or credit) to send the amount specified in the packet. If so, the connector reduces the sender's bandwidth by the value of the packet. If not, the connector returns an ILP Reject packet.
+4. Connector uses the destination ILP address from the packet and its local routing tables to determine which next hop to forward the packet to.
+5. Connector modifies the packet amount and expiry to apply their exchange rate and shorten the expiry time.
+6. Connector forwards the packet to the next hop, which may be another connector. All subsequent connectors go through steps 3-6 until the packet reaches the receiver.
+7. Receiver checks the ILP Prepare packet, according to whatever is stipulated by the higher-level protocol (such as checking whether the amount received is above some minimum specified by the sender). 
+8. Receiver returns an ILP Fulfill packet, containing the preimage of the condition in the ILP Prepare packet. The receiver may include additional data, the formatting of which is determined by the higher-level protocol.
+9. Connector checks that the receiver has returned the fulfillment before the expiry in the ILP Prepare packet. If so, the connector returns the same ILP Fulfill packet to the previous connector and credits the receiver's account with the amount in the ILP Prepare packet. If the Prepare has expired, the connector returns an ILP Reject packet to the previous connector. Each connector repeats this step until the Fulfill (or Reject) packet reaches the sender.
+10. Sender checks that the preimage in the ILP Fulfill packet matches the condition in their Prepare packet and may read any data returned by the receiver.
+11. Sender signs an update to their payment channel with the first connector to cover the value of the ILP Prepare (or they may settle less frequently using slower on-ledger transfers or physical deliveries of cash or other goods) and communicates it to the connector.
+12. Connector validates the sender's signed payment channel update (or other proof of settlement) and increases the sender's bandwidth by the amount they paid. If the sender does not pay for the ILP Prepare, the connector may refuse to process further ILP packets until they do or the connector may block that user entirely.
```

Comment:
This definitely sounds like a bigger risk on connectors than in ILPv1, even if the amounts are smaller. In v1, connectors only had to choose ledgers they trust to enforce cryptographic escrow. Now it sounds like senders can shop around for sucker connectors. Given that this protocol does not define any way to identify senders, blocking them permanently could be quite a challenge.

I suspect that this is supposed to be solved in a higher-level protocol or in the setup of the payment channels, but the phrasing of this point doesn't really make that clear.

Mostly unrelated, this is another place "bandwidth" is used in a somewhat mysterious way.

Reply:
It's certainly a different type of risk but I'm not sure it's a bigger one. In ILPv1, connectors ran the risk that some sender would tie up their liquidity for a long period of time and then either reject the payment or accept it only if the exchange rate had moved in their favor (the free option problem). A big issue with that was that anyone could cause that to happen, even someone that was indirectly connected.

Now, the model is that each bilateral connection requires some level of trust. Either the sender needs to trust the connector (if the account is pre-funded) or the connector needs to trust the sender (if their account is post-funded, as in this example). Like in ILPv1, you need to trust whoever you're directly connected to, but now you're directly connected to other connectors (or senders or receivers), rather than ledgers.